src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Field.java

```diff
@@ -41,22 +42,88 @@ public class Field {
-    genericSignature     = holder.getFieldGenericSignature(fieldIndex);
-
-    name                 = holder.getFieldName(fieldIndex);
+    name = holder.getSymbolFromIndex(values.nameIndex, isInjected());
+    signature = holder.getSymbolFromIndex(values.signatureIndex, isInjected());
+    offset = values.offset;
     id          = new NamedFieldIdentifier(name.asString());
-
-    signature            = holder.getFieldSignature(fieldIndex);
     fieldType   = new FieldType(signature);
+    accessFlags = new AccessFlags(values.accessFlags);
+
+    if (isGeneric()) {
+      genericSignature = holder.getSymbolFromIndex(values.genericSignatureIndex, isInjected());
+    }
+  }
+
+  static class FieldInfoValues {
+    int nameIndex;
+    int signatureIndex;
+    int offset;
+    int accessFlags;
+    int fieldFlags;
+    int initialValueIndex;
+    int genericSignatureIndex;
+    int contendedGroup;
+  }
+
+  // The format of the stream, after decompression, is a series of
+  // integers organized like this:
+  //
+  //   FieldInfo := j=num_java_fields k=num_internal_fields Field*[j+k] End
+  //   Field := name sig offset access internal Optionals(internal)
+  //   Optionals(i) := initval?[i&is_init]     // ConstantValue attr
+  //                   gsig?[i&is_generic]     // signature attr
+  //                   group?[i&is_contended]  // Contended anno (group)
+  //   End = 0
+  //
 
-    short access         = holder.getFieldAccessFlags(fieldIndex);
-    accessFlags = new AccessFlags(access);
+  static FieldInfoValues readFieldInfoValues(CompressedReadStream crs) {
+    FieldInfoValues fieldInfoValues = new FieldInfoValues();
+    fieldInfoValues.nameIndex = crs.readInt(); // read name_index
+    fieldInfoValues.signatureIndex = crs.readInt(); // read signature index
+    fieldInfoValues.offset = crs.readInt(); // read offset
+    fieldInfoValues.accessFlags = crs.readInt(); // read access flags
+    fieldInfoValues.fieldFlags = crs.readInt();  // read internal field flags
+    // Optional reads
+    if (fieldIsInitialized(fieldInfoValues.fieldFlags)) fieldInfoValues.initialValueIndex = crs.readInt(); // read initial value index
+    if (fieldIsGeneric(fieldInfoValues.fieldFlags))     fieldInfoValues.genericSignatureIndex = crs.readInt(); // read generic signature index
+    if (fieldIsContended(fieldInfoValues.fieldFlags))   fieldInfoValues.contendedGroup = crs.readInt(); // read contended group
+    return fieldInfoValues;
   }
 
+  private static FieldInfoValues getFieldInfoValues(CompressedReadStream crs, int fieldIndex) {
+    int javafieldsCount = crs.readInt(); // read num_java_fields
+    int VMFieldsCount = crs.readInt(); // read num_injected_fields;
+    if (fieldIndex < 0 || fieldIndex >= javafieldsCount + VMFieldsCount) throw new IndexOutOfBoundsException();
+    FieldInfoValues fieldInfoValues = null;
+    for (int i = 0; i <= fieldIndex; i++) {
+      fieldInfoValues = readFieldInfoValues(crs);
```

Comment:
Rather than actually creating an unused FieldInfoValues for every field up to the index (which requires reading in the entire stream up to that point), can you just skip over each Field. I think the only FieldInfoValues field you actually need to read in for each Field is fieldFlags so you can compute the presence of optional fields. I think you can use CompressedStream.getPosition() and setPosition() to skip over parts of the stream, although adding CompressedStream.skip(int) would probably be better.

Reply:
The new implementation adds a lazily initialized Field cache to InstanceKlass, so the stream is decoded only once. Not having to decode the stream multiple times makes a skip() method useless.